xla/service/gpu/loop_double_buffer_transformer.cc

```diff
@@ -0,0 +1,405 @@
+    }
+    VLOG(2) << "Peeling instruction " << old_instr->ToString();
+    HloInstruction* new_instr;
+    if (loop_input_consumers.find(old_instr) != loop_input_consumers.end()) {
+      // If instruction is the consumer of loop inputs, then we clone and put it
+      // in the map directly.
+      new_instr = parent_comp->AddInstruction(
+          old_instr->Clone(/*suffix=*/context.suffix()));
+      // There will be cases that other operands of this loop input consumer are
+      // already cloned, replace those operands with corresponding clones.
+      for (int64_t op_ind = 0; op_ind < new_instr->operand_count(); op_ind++) {
+        HloInstruction* op = new_instr->mutable_operand(op_ind);
+        if (old_to_new_map.find(op) != old_to_new_map.end()) {
+          TF_RETURN_IF_ERROR(
+              new_instr->ReplaceOperandWith(op_ind, old_to_new_map[op]));
+        }
+      }
+    } else {
+      // otherwise its operands should have been cloned already since
+      // we traversed the graph in post order. Find its operands
+      // in the map and clone with new operands.
+      std::vector<HloInstruction*> new_operands;
+      for (HloInstruction* old_operand : old_instr->mutable_operands()) {
+        new_operands.push_back(old_to_new_map[old_operand]);
+      }
+      new_instr = parent_comp->AddInstruction(old_instr->CloneWithNewOperands(
+          old_instr->shape(), new_operands, &context));
+    }
+    SetChannelIdForNewCollective(new_instr, module);
+    old_to_new_map[old_instr] = new_instr;
+    all_new_instructions.push_back(new_instr);
+    VLOG(2) << "Added instruction " << new_instr->ToString()
+            << " to parent computation.";
+  }
+  // Handle existing control dependencies.
+  for (HloInstruction* old_instr : all_old_instructions) {
+    if (old_to_new_map.find(old_instr) != old_to_new_map.end()) {
+      HloInstruction* new_instr = old_to_new_map[old_instr];
+      VLOG(2) << "Processing control predecessors for peeled instruction "
+              << new_instr->ToString();
+      if (old_instr->control_predecessors().size() > 0) {
+        std::vector<HloInstruction*> new_control_pred(
+            old_instr->control_predecessors().size());
+        const std::vector<HloInstruction*>& preds =
+            old_instr->control_predecessors();
+        for (int64_t i = 0; i < preds.size(); i++) {
+          new_control_pred[i] = old_to_new_map[preds[i]];
+        }
+
+        TF_RETURN_IF_ERROR(new_instr->DropAllControlDeps());
+        for (HloInstruction* new_pred : new_control_pred) {
+          TF_RETURN_IF_ERROR(new_pred->AddControlDependencyTo(new_instr));
+          VLOG(2) << "Adding " << new_pred->ToString()
+                  << " to control dependency of peeled instruction: "
+                  << new_instr->ToString();
+        }
+      }
+    }
+  }
+
+  // Connect the peeled instructions with previous while loop input
+  // producers.
+  for (auto& instr_index_pair : loop_input_consumers) {
```

Comment:
The main purpose of this is to keep track of boundary instructions so we can inject control dependencies between iterations. loop_input_consumers together with the old_to_new mapping will give all boundary instructions in odd iteration, old_roots will give all boundary instructions in even iteration. then we add `old_roots` as control dep to `old_to_new[loop_input_consumer]`. We still need to know for each operand index of odd-iteration boundary instruction, what output tuple index it should consume, so im just using the struct to store all this info instead of inspecting the instruction.
I removed the tuple creation. The control dep injection loop still needs to be there.

Reply:
Just to check my understanding, currently we do the following:
1. clone all instructions with corresponding operands
2. for boundary instructions, their original operands are get_tuple_element which indexes into the input tuple, we dont want to clone the GTEs, so we want to directly connect the producers in the original output tuple index to the boundary instructions that consume the same input tuple index. Therefore I pre-process the original input consumers and input GTE indices and store the mapping in LoopInputConsumer. Without this, when we clone instructions, we still need to bookkeep all boundary instructions and see if current instruction is a boundary, if so, we will need to get the index from the GTE and find the correct operand in the original output tuple.
3. inject control deps for boundary instructions.

 I think 2 is where you are suggesting to just clone the tuple and GTEs and let TupleSimplifier to take care of it? It looks like we are creating something that we don't need and relying on another pass to do the right thing.

Reply:
That's pretty much it, although I will note that I only suggest creating the tuple for the sake of simplicity, because it avoids fancier analysis. It's also possible to avoid doing that, still without requiring any bookkeeping. To be clear, what I suggest is not substantially different, but I think it makes the code easier to follow.

For the sake of the explanation, let's consider only the peeled loop iteration (it works similarly inside the loop body). Let's say the loop looks roughly like this:

```
body {
  input_tuple_in_loop = (tuple_arg0, tuple_arg1)
  arg0 = gte(input_tuple, 0)
  arg1 = gte(input_tuple, 1)
  ...
  root_tuple = ...
}

entry {
  value0 = ...
  value1 = ...
  input_tuple = (value0, value1)
  while(tuple), body=body
}
```

With tuples, we do as follows:

1. populate `old_to_new` with `input_tuple_in_loop -> input_tuple`
2. copy all instructions not in the map in post-order
3. replace the parameter to the while loop with `old_to_new[root_tuple]`
(no need for control dependencies outside of the loop, I don't think?)

Without tuples, we do as follows:

1. traverse all users of `input_tuple_in_loop`, and assuming they are all GTEs, populate `old_to_new` with a map from each GTE to the corresponding tuple element in `input_tuple`, i.e. `arg0 -> value0`, `arg1 -> value1`; add a dummy mapping for `input_tuple_in_loop` to avoid processing it in 2.
2. copy all instructions not in the map in post-order
3. replace the parameter to the while loop with `old_to_new[root_tuple]` (you need that tuple anyway)

Step 1. in the "without tuples" example is very close to what you are doing, but it doesn't require preprocessing the input. Rather, it just directly maps each instruction to what it needs to be at cloning time.

I like the "with tuples" example because it is more straightforward (as you can see from above, the steps are easier to describe, and do not need much special casing).

> I think 2 is where you are suggesting to just clone the tuple and GTEs and let TupleSimplifier to take care of it? It looks like we are creating something that we don't need and relying on another pass to do the right thing.

Sure, but this is pretty much in the same way as we are relying on DCE to eliminate new dead instructions. I'm not opposed to using an approach that does not materialise tuples if you prefer it, but I still think that we can simplify the code using a process like the one described above.

Similarly, we could also remove the newly dead instructions directly here, and avoid calling DCE outside of the pass.

Reply:
Thanks for the explanation, let me take a stab at the one with tuples, i think that should be quicker to implement.

Reply:
Hmm I'm getting this error when cloning:
`Check failed: instruction->opcode() != HloOpcode::kParameter Parameter instructions cannot be added to a computation after it has been built`
So i guess cloning tuple is not an option

Reply:
@bchetioui  I went for the tuples approach, and it did simplify the copy loop quit a bit. Please take another look. Thanks.